### PR TITLE
Add electrum support

### PR DESCRIPTION
**Pull request by dr-orlovsky**
These are electrum-specific commits extracted from `v0.11` branch https://github.com/RGB-WG/rgb/pull/103, including the commit from https://github.com/RGB-WG/rgb/pull/126 PR, which I move here to be able to run CI on the whole thing.

We need to make sure this compiles to WASM before merging - and I expect it should due to https://github.com/BP-WG/bp-wallet/pull/15/commits/9d1cf14c1a3c6810827f1de9b887103f3f178ac8, making https://github.com/RGB-WG/rgb/pull/127 and https://github.com/RGB-WG/rgb/pull/128 outdated.

I will wait for @crisdut and @zoedberg confirmation before merging this and close the other two PRs trying to fix default features of `bp-wallet`

### CODE REVIEW

**Comment by zoedberg** [comment]
it works good, thanks

**Review by zoedberg — APPROVED** [review]

**Comment by crisdut** [comment]
I tried compile for wasm target, but I receive the error:

`wasm-pack build --no-default-features --features serde`

```
error[E0004]: non-exhaustive patterns: type `&AnyResolver` is non-empty
  --> src/resolvers/any.rs:77:15
   |
77 |         match self {
   |               ^^^^
   |
note: `AnyResolver` defined here
  --> src/resolvers/any.rs:34:10
   |
34 | pub enum AnyResolver {
   |          ^^^^^^^^^^^
   = note: the matched value is of type `&AnyResolver`
   = note: references are always considered inhabited
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern as shown
   |
77 ~         match self {
78 +             _ => todo!(),
79 +         }
   |
```

We Suppress AnyResolver when we compile for `wasm`. Given that [blocking http calls](https://github.com/rustwasm/wasm-bindgen/issues/2111), don't work when working with wasm, it wouldn't make sense to expose this solution.

We can add something similar to the prefetches we use in `bitmask-core` in the future, something like AnyResolver::InMemory, if that makes sense.

**Comment by dr-orlovsky** [comment]
Yeah, that's why I suggested making `AnyResolver` non-exhaustive here: https://github.com/RGB-WG/rgb/pull/126#issuecomment-1936948139

Also, this is the reason I was against this approach: https://github.com/RGB-WG/rgb/pull/126#pullrequestreview-1874974590

I do not remember all the reasons, but when we tried it in `rust-bitcoin`, it was a bug of pain in all downstream dependencies, so I wonder if we should go this route...

**Comment by zoedberg** [comment]
By running the command that @crisdut is giving (`wasm-pack build --no-default-features --features serde`) I receive the following:
```
Error: crate-type must be cdylib to compile to wasm32-unknown-unknown. Add the following to your Cargo.toml file:

[lib]
crate-type = ["cdylib", "rlib"]
Caused by: crate-type must be cdylib to compile to wasm32-unknown-unknown. Add the following to your Cargo.toml file:

[lib]
crate-type = ["cdylib", "rlib"]
```

By adding the `crate-type` as suggested I receive the following errors:
```
error[E0599]: the method `fill` exists for struct `SystemRandom`, but its trait bounds were not satisfied
   --> /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustls-0.21.10/src/rand.rs:8:10
    |
7   | /     SystemRandom::new()
8   | |         .fill(bytes)
    | |         -^^^^ method cannot be called on `SystemRandom` due to unsatisfied trait bounds
    | |_________|
    | 
    |
   ::: /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ring-0.17.7/src/rand.rs:112:1
    |
112 |   pub struct SystemRandom(());
    |   -----------------------
    |   |
    |   doesn't satisfy `SystemRandom: SecureRandom`
    |   doesn't satisfy `SystemRandom: ring::rand::sealed::SecureRandom`
    |
    = note: the following trait bounds were not satisfied:
            `SystemRandom: ring::rand::sealed::SecureRandom`
            which is required by `SystemRandom: SecureRandom`

error[E0277]: the trait bound `SystemRandom: ring::rand::sealed::SecureRandom` is not satisfied
  --> /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustls-0.21.10/src/kx.rs:32:86
   |
32 |             ring::agreement::EphemeralPrivateKey::generate(skxg.agreement_algorithm, &rng).ok()?;
   |                                                                                      ^^^^ the trait `ring::rand::sealed::SecureRandom` is not implemented for `SystemRandom`
   |
   = help: the following other types implement trait `ring::rand::sealed::SecureRandom`:
             FixedByteRandom
             FixedSliceRandom<'_>
             FixedSliceSequenceRandom<'_>
   = note: required for `SystemRandom` to implement `SecureRandom`
   = note: required for the cast from `&SystemRandom` to `&dyn SecureRandom`

error[E0277]: the trait bound `SystemRandom: ring::rand::sealed::SecureRandom` is not satisfied
   --> /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustls-0.21.10/src/sign.rs:196:34
    |
196 |             .sign(self.encoding, &rng, message, &mut sig)
    |                                  ^^^^ the trait `ring::rand::sealed::SecureRandom` is not implemented for `SystemRandom`
    |
    = help: the following other types implement trait `ring::rand::sealed::SecureRandom`:
              FixedByteRandom
              FixedSliceRandom<'_>
              FixedSliceSequenceRandom<'_>
    = note: required for `SystemRandom` to implement `SecureRandom`
    = note: required for the cast from `&SystemRandom` to `&dyn SecureRandom`

error[E0277]: the trait bound `SystemRandom: ring::rand::sealed::SecureRandom` is not satisfied
   --> /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustls-0.21.10/src/sign.rs:232:50
    |
232 |         EcdsaKeyPair::from_pkcs8(sigalg, &der.0, &rng)
    |                                                  ^^^^ the trait `ring::rand::sealed::SecureRandom` is not implemented for `SystemRandom`
    |
    = help: the following other types implement trait `ring::rand::sealed::SecureRandom`:
              FixedByteRandom
              FixedSliceRandom<'_>
              FixedSliceSequenceRandom<'_>
    = note: required for `SystemRandom` to implement `SecureRandom`
    = note: required for the cast from `&SystemRandom` to `&dyn SecureRandom`

error[E0277]: the trait bound `SystemRandom: ring::rand::sealed::SecureRandom` is not satisfied
   --> /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustls-0.21.10/src/sign.rs:234:78
    |
234 |             .or_else(|_| Self::convert_sec1_to_pkcs8(scheme, sigalg, &der.0, &rng))
    |                                                                              ^^^^ the trait `ring::rand::sealed::SecureRandom` is not implemented for `SystemRandom`
    |
    = help: the following other types implement trait `ring::rand::sealed::SecureRandom`:
              FixedByteRandom
              FixedSliceRandom<'_>
              FixedSliceSequenceRandom<'_>
    = note: required for `SystemRandom` to implement `SecureRandom`
    = note: required for the cast from `&SystemRandom` to `&dyn SecureRandom`

error[E0277]: the trait bound `SystemRandom: ring::rand::sealed::SecureRandom` is not satisfied
   --> /home/zoe/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rustls-0.21.10/src/sign.rs:317:19
    |
317 |             .sign(&rng, message)
    |                   ^^^^ the trait `ring::rand::sealed::SecureRandom` is not implemented for `SystemRandom`
    |
    = help: the following other types implement trait `ring::rand::sealed::SecureRandom`:
              FixedByteRandom
              FixedSliceRandom<'_>
              FixedSliceSequenceRandom<'_>
    = note: required for `SystemRandom` to implement `SecureRandom`
    = note: required for the cast from `&SystemRandom` to `&dyn SecureRandom`
```

so maybe to build on wasm we need to patch some dependencies differently?

Anyway, when implementing the `AnyResolver` I was not considering that someone could need building the library with nor electrum nor esplora features. The fix for this should be simple, just protect the `any` module and the `AnyResolver` like so:
```rust
#[cfg(feature = "electrum")]
#[cfg(feature = "esplora_blocking")]
mod any;

#[cfg(feature = "electrum")]
#[cfg(feature = "esplora_blocking")]
pub use any::AnyResolver;
```

**Comment by cryptoquick** [comment]
I think I figured out what was keeping this from compiling to WASM. See #132 and https://github.com/BP-WG/bp-wallet/pull/16

File: Cargo.toml [hunk 1]
```diff
@@ -66,13 +68,15 @@ name = "rgb_rt"
 [dependencies]
 amplify = { workspace = true }
 baid58 = { workspace = true }
+bitcoin = { workspace = true, optional = true }
 commit_verify = { workspace = true }
 strict_types = { workspace = true }
 bp-core = { workspace = true }
 bp-std = { workspace = true }
 bp-wallet = { workspace = true, features = ["fs"] }
```
[the remark this answers is not in the document]
Comment (zoedberg) [reply Cargo.toml:76]: Hi @crisdut, sorry for the late answer but I've somehow missed this. Thanks for the tip, unfortunately though I'm still receiving the previously reported errors.

**Comment by zoedberg** [comment]
> I think I figured out what was keeping this from compiling to WASM. See #132

Nice, so it seems #132, which is using my suggested fix, has fixed the wasm compilation (I've tried the branch and I can confirm it). So @dr-orlovsky, can we merge that PR into this and then this PR into master?

**Pull request merged**